Add time tracking plugin

takuyakanbr opened the pull request:
**Changes**

- Convert farming tracker to time tracking plugin
  - Flower, hops, and grape patches now have their own tabs
- Add bird house tracker and a new bird house tab
  - The plugin can optionally send a notification when all bird houses are ready to be dismantled
- Add a new clock tab that contains user-defined timers and stopwatches
  - Timers and stopwatches continue 'running' when client is closed
  - Timers can optionally send a notification when they finish counting down
- Add a new overview tab that provides a quick summary of the various tabs
- Add tooltips to tab icons

**Screenshots**

Plugin icon
![image0](https://user-images.githubusercontent.com/875036/43277468-bb49eb90-913a-11e8-981b-666bf6d8f10b.png)

Overview tab
![overview](https://user-images.githubusercontent.com/875036/43349681-1baeac38-9234-11e8-93d6-d066b6986d3b.png)

Timers and stopwatches
![image1](https://user-images.githubusercontent.com/875036/43330553-19ce77ba-91f6-11e8-9187-55832e45056e.png)

Bird house tracker
![image2](https://user-images.githubusercontent.com/875036/43330555-1c449966-91f6-11e8-8fa7-b0ceed824def.png)

**Notes**

All existing farming tracker data will be migrated over to the new plugin. No data is expected to be lost.

I would suggest that the migration code (`FarmingTracker#migrateConfiguration`) should be removed after around 2 to 4 weeks.

Closes #259
Closes #3581
Fixes #1224
Fixes #1612

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/Tab.java:
```diff
@@ -32,12 +32,17 @@
+	GRAPE("Grape Patches", ItemID.GRAPES),
+	SPECIAL("Special Patches", ItemID.MUSHROOM);
+
+	public static final Tab[] FARMING_TABS = {ALLOTMENT, HERB, TREE, FRUIT_TREE, FLOWER, HOPS, BUSH, GRAPE, SPECIAL};
```
Prefer using `Tab.values()` instead

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/hunter/BirdHouse.java:
```diff
@@ -0,0 +1,67 @@
+	MAGIC("Magic Bird House", ItemID.MAGIC_BIRD_HOUSE),
+	REDWOOD("Redwood Bird House", ItemID.REDWOOD_BIRD_HOUSE);
+
+	static final BirdHouse[] BIRD_HOUSES = {NORMAL, OAK, WILLOW, TEAK, MAPLE, MAHOGANY, YEW, MAGIC, REDWOOD};
```
`Birdhouse.values()`

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/TimeTrackingPlugin.java:
```diff
@@ -108,18 +121,44 @@ protected void startUp() throws Exception
+			@Override
+			public void run()
+			{
+				updatePanel();
```
We do have a `@Schedule` (https://static.runelite.net/api/runelite-client/net/runelite/client/task/Schedule.html) although I do not think it provides sub-tick accuracy currently. I would either maybe modify that, or at least inject the scheduler and schedule it with that. Each java `Timer` is always backed by a java `Thread` which is not great.

Updating swing panels often is a performance issue and even now we have users report that when the farming panel is open (thus, when it updates) it causes noticeable performance issues. Can we make this not update if there are no timesr running? Or if it isn't open?

takuyakanbr replied:
Even if I decrease the tick interval of the `Scheduler`, I can't currently use it because `Hooks#clientMainLoop` does not run when idling on the login screen. 

Only 1 `Timer` is created and used when the plugin is active, which imo is insignificant in terms of resource consumption. (If and when another plugin requires a timer, it can perhaps be moved into `Scheduler` or another class.)

None of the tabs and their contents are being updated when the tab isn't open. (Previously, all patch panels were being updated when updating the UI.) I have also updated it such that while on the clock tab, no UI updates are being done if there are no active clocks.

Adam- replied:
I would maybe prefer if you left the scheduler alone and scheduled it on the `ScheduledExecutorService` which you can inject.

Abextm commented:
- Partially seeded birdhouses varp%3==2 should probably show as BUILT, not UNKNOWN
 - Birdhouses should probably use the same color code as farming patches to be consistent
 - When I was messing with birdhouses they did tick similar to farming patches, but at a relatively fast rate, so simply adding 50 minutes to the setup time was not entirely accurate, but I'm not sure if there is anything we can really do about that. 
 - Logging in on fossil island resets the birdhouse times (this bug also applies to farming patches, but it doesn't really matter because they just get fixed when you finish logging in). The best way to fix this I think would be to make sure the Click here to continue screen widget isn't showing, and wasn't showing the previous tick. It would also be nice to add the plane check to farming patches because they need it too.
 - The timer UI is rather clunky, perhaps it would be better to make the shown time a input. If this isn't possible I would still like to have something that doesn't open a dialog box
 - Same thing with rename, it could just be a text field that activates on focus without being as modal

takuyakanbr commented:
Added an overview tab (see first comment for screenshot) and addressed above comments.

deathbeam commented:
I would put birdhouses before farming as its single-entry compared to x amount of different entries for farming

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/clocks/ClockTabPanel.java:
```diff
@@ -0,0 +1,235 @@
+				resetIcon = ImageIO.read(TimeTrackingPlugin.class.getResourceAsStream("reset_icon.png"));
+				startIcon = ImageIO.read(TimeTrackingPlugin.class.getResourceAsStream("start_icon.png"));
+				ADD_ICON = new ImageIcon(ImageIO.read(TimeTrackingPlugin.class.getResourceAsStream("add_icon.png")));
+				ADD_ICON_HOVER = new ImageIcon(ImageIO.read(TimeTrackingPlugin.class.getResourceAsStream("add_hover_icon.png")));
```
What's the reason for two separate images here instead of using `SwingUtil::grayscaleOffset` like the other hovered icon images?

takuyakanbr commented:
Added notifications for bird house tracker and farming tracker. (Most of the necessary infrastructure was already added to support the overview tab anyway.) Also addressed all pending comments.

For the bird house tracker, I am now blocking transitions from 3 or 4 non-empty bird houses to 4 empty bird houses, since it seems that the client might occasionally not receive the varps in a timely manner.

The code should be ready for reviewing now.

Abextm commented:
- There is definitely something wrong with the prediction now, my trees which should be done on Sunday are saying Thursday, thought their progress bar is correct. My birdhouses I just planted (on Saturday) say they will be done on Wednesday.
 - My tree patches say ready, even though my hardwoods are not (spirit trees are all done, and normal trees are not planted)
 - Bird houses still can reset during the login process, the issue is not that the server doesn't send the varps in a consistent time, it is that it only sends them the tick after you exit the "click here to continue" splash screen
 - There doesn't seem to be any feedback when a timer expires, which makes them somewhat useless imo
 - Timers should automatically reset when you start them.

takuyakanbr commented:
Updated to resolve the above issues.

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/hunter/BirdHouseData.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Contains data about the state of a particular {@link BirdHouseSpace}, at a particular point in time.
+ */
+@AllArgsConstructor
```
`@AllArgsConstructor` is redundant here

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/TimeTrackingPlugin.java:
```diff
@@ -180,13 +222,19 @@ public void checkCompletion()
-		SwingUtilities.invokeLater(panel::update);
+		long unitTime = Instant.now().toEpochMilli() / 200;
+
+		boolean clockDataChanged = unitTime % 5 == 0 && clockManager.checkCompletion();
```
Non functional methods should not be called with short circuiting operators.

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/clocks/ClockTabPanel.java:
```diff
@@ -0,0 +1,239 @@
+		removeAll();
+		clockPanels.clear();
+
+		GridBagConstraints c = new GridBagConstraints();
```
DynamicGridLayout is preferred if possible

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingTracker.java:
```diff
@@ -56,6 +58,15 @@
+	 * Each value is set to {@code 0} if all patches of that type have already completed
+	 * when updating the value.
+	 */
+	private Map<PatchImplementation, Long> completionTimes = new HashMap<>();
```
EnumMap

takuyakanbr commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/hunter/BirdHouseTracker.java:
```diff
@@ -0,0 +1,210 @@
+
+	public BirdHouseTabPanel createBirdHouseTabPanel()
+	{
+		return new BirdHouseTabPanel(itemManager, this, config);
```
That will cause the UI to screw up. Looks like guice creates it before the UI theme has been set?
![a1](https://user-images.githubusercontent.com/875036/43372854-02dcdc8e-93d8-11e8-99fb-ef1b679509a7.png)

takuyakanbr commented:
Made requested changes.

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/hunter/BirdHouseTracker.java:
```diff
@@ -0,0 +1,223 @@
+	private Notifier notifier;
+
+	@Getter(AccessLevel.PACKAGE)
+	private final Map<BirdHouseSpace, BirdHouseData> birdHouseData = new HashMap<>();
```
This is accessed from multiple threads

takuyakanbr replied:
Switched to `ConcurrentHashMap`

Abextm replied:
Its still not thread safe because you iterate over it

takuyakanbr replied:
All updates and iterations happen on the same (client) thread though. The UI thread uses only `Map#get`.

deathbeam commented:
Needs rebase

deathbeam commented:
Also, the grayscaleOffset etc methods were moved to ImageUtil, so that needs to be updated here too.

takuyakanbr commented:
Made requested changes.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/TimeTrackingPlugin.java:
```diff
@@ -0,0 +1,240 @@
+	@Override
+	protected void shutDown() throws Exception
+	{
+		lastTickLocation = null;
```
set lastTickPostLogin to false here too

takuyakanbr replied:
Updated

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/timetracking/clocks/ClockTabPanel.java:
```diff
@@ -0,0 +1,191 @@
+		BufferedImage addIcon = ImageUtil.getResourceStreamFromClass(TimeTrackingPlugin.class, "add_icon.png");
+
+		DELETE_ICON = new ImageIcon(deleteIcon);
+		DELETE_ICON_HOVER = new ImageIcon(ImageUtil.grayscaleOffset(deleteIcon, -100));
```
The behavior of `grayscaleOffset` has changed somewhat due to my PR. Double-check to see if these should still be using this method (which now explicitly reduces RGB values, *and not alpha*) or if they should use `alphaOffset` instead.
